libcylinder/src/key/load.rs

```diff
@@ -0,0 +1,324 @@
+            Err(KeyLoadError(format!(
+                "Failed to load key: could not be found {}",
+                path.as_path().display()
+            )))
+        }
+    }
+}
+
+// Reads a private key from the given path
+fn read_private_key(path: PathBuf) -> Result<PrivateKey, KeyLoadError> {
+    let mut file = File::open(&path).map_err(|err| {
+        KeyLoadError(format!(
+            "Unable to open key file '{}': {}",
+            path.display(),
+            msg_from_io_error(err)
+        ))
+    })?;
+
+    let mut buf = String::new();
+    file.read_to_string(&mut buf).map_err(|err| {
+        KeyLoadError(format!(
+            "Unable to read key file '{}': {}",
+            path.display(),
+            msg_from_io_error(err)
+        ))
+    })?;
+    // let key = buf.trim().to_string();
+    let key = match buf.lines().next() {
+        Some(k) => k.trim().to_string(),
+        None => {
+            return Err(KeyLoadError(format!("Empty key file: {}", path.display())));
+        }
+    };
+
+    Ok(PrivateKey::new_from_hex(&key)
+        .map_err(|err| KeyLoadError(format!("unable to create private key from hex: {}", err)))?)
+}
+
+fn msg_from_io_error(err: IoError) -> String {
+    match err.kind() {
+        ErrorKind::NotFound => "File not found".into(),
+        ErrorKind::PermissionDenied => "Permission denied".into(),
+        ErrorKind::InvalidData => "Invalid data".into(),
+        _ => "Unknown I/O error".into(),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::PrivateKey;
+
+    use std::fs::File;
+    use std::io::Write;
+    use tempdir::TempDir;
+
+    // tests that when an existing key name and default path are given load_user_key returns the
+    // private key
+    #[test]
+    fn retrieve_key_success() {
+        let temp_dir = TempDir::new("test_key_dir").expect("Failed to create temp dir");
+
+        let key_name = "test_key.priv";
+        let default_path = temp_dir
+            .path()
+            .to_str()
+            .expect("Failed to get default path");
+
+        let key_path = temp_dir
+            .path()
+            .join(key_name)
+            .to_str()
+            .expect("Failed to get path")
+            .to_string();
+
+        let mut temp_file =
+            File::create(&key_path).expect("Unable to create temp private key file");
+
+        let private_key = PrivateKey::new(vec![
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 1,
+        ]);
+
+        writeln!(temp_file, "{}", private_key.as_hex())
+            .expect("Unable to write private key to file");
+
+        let retrieved_private_key =
+            load_user_key(Some(key_name), default_path).expect("Unable retrieve key from file");
+
+        assert_eq!(retrieved_private_key.into_bytes(), private_key.into_bytes(),);
+    }
+    // tests that when no key name is given and the USER environment variable is set the value of
```

Comment:
Insert a line above this (and same for instances below).

Reply:
Fixed